openedx/core/djangoapps/schedules/resolvers.py

```diff
@@ -1,136 +1,313 @@
+            'courseenrollment__schedule__{}__lt'.format(self.schedule_date_field): target_day + datetime.timedelta(days=1),
+        }
+        users = User.objects.filter(
+            courseenrollment__is_active=True,
+            **schedule_day_equals_target_day_filter
+        ).annotate(
+            id_mod=F('id') % self.num_bins
+        ).filter(
+            id_mod=self.bin_num
+        )
+
+        schedule_day_equals_target_day_filter = {
+            '{}__gte'.format(self.schedule_date_field): target_day,
+            '{}__lt'.format(self.schedule_date_field): target_day + datetime.timedelta(days=1),
+        }
+        schedules = Schedule.objects.select_related(
+            'enrollment__user__profile',
+            'enrollment__course',
+        ).prefetch_related(
+            'enrollment__course__modes'
+        ).filter(
+            Q(enrollment__course__end__isnull=True) | Q(
+                enrollment__course__end__gte=self.current_datetime),
+            enrollment__user__in=users,
+            enrollment__is_active=True,
+            **schedule_day_equals_target_day_filter
+        ).order_by(order_by)
+
+        if self.org_list is not None:
+            if self.exclude_orgs:
+                schedules = schedules.exclude(enrollment__course__org__in=self.org_list)
+            else:
+                schedules = schedules.filter(enrollment__course__org__in=self.org_list)
+
+        if "read_replica" in settings.DATABASES:
+            schedules = schedules.using("read_replica")
+
+        LOG.debug('Query = %r', schedules.query.sql_with_params())
+
+        with function_trace('schedule_query_set_evaluation'):
+            # This will run the query and cache all of the results in memory.
+            num_schedules = len(schedules)
+
+        # This should give us a sense of the volume of data being processed by each task.
+        set_custom_metric('num_schedules', num_schedules)
+
+        return schedules
 
 
 class ScheduleStartResolver(BinnedSchedulesBaseResolver):
     """
     Send a message to all users whose schedule started at ``self.current_date`` + ``day_offset``.
     """
-    async_send_task = recurring_nudge_schedule_bin
+    log_prefix = 'Scheduled Nudge'
+    schedule_date_field = 'start'
     num_bins = RECURRING_NUDGE_NUM_BINS
-    enqueue_config_var = 'enqueue_recurring_nudge'
 
-    def __init__(self, *args, **kwargs):
-        super(ScheduleStartResolver, self).__init__(*args, **kwargs)
-        self.log_prefix = 'Scheduled Nudge'
+    def schedules_for_bin(self):
+
+        schedules = self.get_schedules_with_target_date_by_bin_and_orgs()
+
+        for (user, user_schedules) in groupby(schedules, lambda s: s.enrollment.user):
+            user_schedules = list(user_schedules)
+            course_id_strs = [str(schedule.enrollment.course_id) for schedule in user_schedules]
+
+            first_schedule = user_schedules[0]
+            template_context = get_base_template_context(self.site)
```

Comment:
This can be moved out of the `for` loop.  It may not be that costly to compute - but it's anyway not needed to recompute per user.

Reply:
Good call.

Reply:
done.